File: rl_python/match3_engine/board.py
from __future__ import annotations

from typing import List, Optional, Tuple

from .cells import Cell, clone_cell
from .constants import ROWS, COLS


Board = List[List[Optional[Cell]]]


def cell_key(r: int, c: int) -> str:
    return f"{r},{c}"


def create_empty_board() -> Board:
    return [[None for _ in range(COLS)] for _ in range(ROWS)]


def clone_board(board: Board) -> Board:
    return [[clone_cell(cell) for cell in row] for row in board]


def in_bounds(r: int, c: int) -> bool:
    return 0 <= r < ROWS and 0 <= c < COLS


def cells_equal(a: Optional[Cell], b: Optional[Cell]) -> bool:
    if a is None and b is None:
        return True
    if a is None or b is None:
        return False
    if a.kind != b.kind or a.shape != b.shape or a.level != b.level:
        return False
    if bool(a.frozen) != bool(b.frozen):
        return False
    a_pt = getattr(a, "powerup_type", "")
    b_pt = getattr(b, "powerup_type", "")
    return a_pt == b_pt

File: src/core/board.js
import { ROWS, COLS } from './constants.js';

export function cellKey(r, c) {
  return `${r},${c}`;
}

export function createEmptyBoard() {
  return Array.from({ length: ROWS }, () => Array(COLS).fill(null));
}

export function cloneBoard(board) {
  return board.map((row) => row.map((c) => (c ? { ...c } : null)));
}

export function inBounds(r, c) {
  return r >= 0 && r < ROWS && c >= 0 && c < COLS;
}

/** 统计棋盘上冰冻格数量（用于解冻任务计数：操作前后差值=本步解冻数）。 */
export function countFrozen(board) {
  let n = 0;
  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      const cell = board[r][c];
      if (cell && cell.frozen) n += 1;
    }
  }
  return n;
}

export function cellsEqual(a, b) {
  if (!a && !b) return true;
  if (!a || !b) return false;
  return (
    a.kind === b.kind &&
    a.shape === b.shape &&
    a.level === b.level &&
    Boolean(a.frozen) === Boolean(b.frozen) &&
    (a.powerupType || '') === (b.powerupType || '')
  );
}

export function boardDiffChangedCells(prevBoard, nextBoard) {
  const changed = [];
  for (let r = 0; r < ROWS; r++) {
    for (let c = 0; c < COLS; c++) {
      if (!cellsEqual(prevBoard[r][c], nextBoard[r][c])) changed.push({ r, c });
    }
  }
  return changed;
}
